src/Npgsql/NpgsqlTypes/NpgsqlDateTime.cs

```diff
@@ -154,7 +154,19 @@ public NpgsqlDateTime ToUniversalTime()
                     return new NpgsqlDateTime(Subtract(TimeZoneInfo.Local.GetUtcOffset(new DateTime(ToDateTime().Ticks, DateTimeKind.Local))).Ticks, DateTimeKind.Utc);
                 }
                 // Else there are no DST rules available in the system for outside the DateTime range, so just use the base offset
-                return new NpgsqlDateTime(Subtract(TimeZoneInfo.Local.BaseUtcOffset).Ticks, DateTimeKind.Utc);
+                var timeTicks = _time.Ticks - TimeZoneInfo.Local.BaseUtcOffset.Ticks;
+                var date = _date;
+                if (timeTicks < 0)
+                {
+                    timeTicks += NpgsqlTimeSpan.TicksPerDay;
+                    date = date.AddDays(-1);
+                }
+                else if (timeTicks > NpgsqlTimeSpan.TicksPerDay)
+                {
+                    timeTicks -= NpgsqlTimeSpan.TicksPerDay;
+                    date = date.AddDays(1);
+                }
+                return new NpgsqlDateTime(date, TimeSpan.FromTicks(timeTicks), DateTimeKind.Utc);
```

Comment:
Can you briefly explain how this fixes things? The previous code which uses the ticks-based constructor does look OK, unless of course that constructor doesn't work in some way...

In any case, this is exactly the kind of date/time logic we don't want to implement in Npgsql, and why we have the conversation about removing or stripping the date/time types (which @YohDeadfall was working on in #3537). I'm not sure we should start fixing bugs in this area now...

Reply:
OK. Do you really think we should work on this given that we're deprecating/replacing these types?

Reply:
That's a tough question. Considering that we're removing them only in a future version, it makes sense to continue supporting them (fixing bugs), but without adding any new functionality.